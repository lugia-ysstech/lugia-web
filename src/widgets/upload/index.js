/*
 *create by LYQ
 *
 *2018-12-03
 *
 *@flow
 *
 */
import Upload from './upload';
import Widget from '../consts/index';
import ThemeHoc from '@lugia/theme-hoc';
import ValidateHoc from '../input/validateHoc';

export default ThemeHoc(ValidateHoc(Upload), Widget.Upload, {
  hover: true,
  active: false,
});
