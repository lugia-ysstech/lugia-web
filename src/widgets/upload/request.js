/*
 *create by LYQ
 *
 *2018-12-05
 *
 *@flow
 *
 */

export function getRequestXHR(): Object {
  return window.XMLHttpRequest
    ? new XMLHttpRequest()
    : new window.ActiveXobject('Microsoft.XMLHTTP');
}

export function getFormData(data: Object, file: Object, uploadFileName: string): Object {
  const newData: Object = new FormData();
  if (data) {
    for (const field in data) {
      newData.append(field, data[field]);
    }
  }
  if (file) newData.append(uploadFileName, file);
  return newData;
}

export function getQueryString(data: ?Object): string {
  if (!data) return '';

  const result = [];

  for (const field in data) {
    result.push(`${field}=${data[field]}`);
  }

  return result.join('&');
}

export function addEventListener(
  target: Object,
  event: string,
  func: Function,
  useCapture?: boolean = false
) {
  target.addEventListener(event, func, useCapture);
}

const doGet = function(xhr, url, data, asynch) {
  const queryString = getQueryString(data);
  xhr.open('get', url + (queryString ? `?${queryString}` : queryString), asynch);
  xhr.send();
};

const doPost = function(xhr, url, { data, headers, file, uploadFileName }, asynch) {
  const params = getFormData(data, file, uploadFileName);
  xhr.open('post', url, asynch);

  if (headers) {
    Object.keys(headers).forEach(field => {
      xhr.setRequestHeader(field, headers[field]);
    });
  }
  xhr.send(params);
};

function parseResponse(xhr, dataType) {
  const type = dataType.toLocaleLowerCase();

  const { responseText, responseXML } = xhr;
  let res;
  switch (type) {
    case 'text':
      res = responseText;
      break;
    case 'xml':
      res = responseXML;
      break;
    case 'json':
      res = JSON.parse(responseText);
      break;
    default:
  }
  return res;
}

function request(dataObject: Object) {
  const { url } = dataObject;
  if (!url) {
    return;
  }

  const xhr = getRequestXHR();

  const { withCredentials = false } = dataObject;
  xhr.withCredentials = withCredentials;

  const { onProgress, onComplete } = dataObject;

  xhr.upload.onprogress = onProgress;

  if (onProgress) {
    addEventListener(xhr.upload, 'progress', onProgress, false);
  }

  if (onComplete) {
    addEventListener(xhr, 'load', onComplete, false);
  }

  const { method = 'get', asynch = true, data, uploadFileName } = dataObject;
  const type = method.toLocaleLowerCase();
  switch (type) {
    case 'get':
      doGet(xhr, url, data, asynch);
      break;
    case 'post':
    default:
      const { headers, file } = dataObject;
      doPost(xhr, url, { data, file, headers, uploadFileName }, asynch);
      break;
  }

  const { onFail, dataType = 'text', onSuccess } = dataObject;
  xhr.onreadystatechange = function() {
    const { readyState } = xhr;
    if (readyState === 4) {
      const { status } = xhr;
      if (status === 200) {
        const res = parseResponse(xhr, dataType);
        onSuccess && onSuccess(res);
      } else {
        const { responseText } = xhr;
        onFail && onFail(responseText);
      }
    }
  };
  return xhr;
}

export default request;
