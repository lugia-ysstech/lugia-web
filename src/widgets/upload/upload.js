/*
 *create by LYQ
 *
 *2018-12-03
 *
 *@flow
 *
 */
import React from 'react';
import GetElement from './getelement';
import request from './request';
import CSSComponent, { css } from '../theme/CSSProvider';

const Container = CSSComponent({
  tag: 'div',
  className: 'upload_Container',
  css: css`
    position: relative;
    display: inline-block;
  `,
});

type UploadProps = {
  disabled?: boolean,
  limit?: number,
  data?: Object,
  areaType?: string,
  multiple?: boolean,
  fileList?: Array<Object>,
  showFileList?: boolean,
  withCredentials?: boolean,
  autoUpload?: boolean,
  accept?: string,
  url: string,
  headers?: Object,
  inputId?: string,
  size?: string,
  accessKey?: Array<string>,
  method?: string,
  dataType?: string,
  beforeUpload?: Function,
  onProgress?: Function,
  onSuccess?: Function,
  onComplete?: Function,
  onChange?: Function,
  onFail?: Function,
  themeProps: Object,
  getPartOfThemeHocProps: Function,
  getPartOfThemeProps: Function,
  defaultTips?: Object,
  userDefine?: any,
};
type StateProps = {
  defaultText?: string,
  classNameStatus?: string,
  fileListDone: Array<Object>,
  isAllowUpload: boolean,
  choosedFile?: Array<Object>,
  previewUrl?: string,
};

export const getIndexInArray = (data: Array<string>, key: string): number => {
  if (!key) return -1;
  return data.indexOf(key);
};

export const isKeyInArray = (data: Array<string>, key: string): boolean => {
  return getIndexInArray(data, key) !== -1;
};

export const isEmptyObject = (obj: Object): boolean => {
  if (!obj) return false;
  return Object.keys(obj).length === 0;
};

export const isIdInArray = (hashMark: string, array: Array<Object>) => {
  if (!array.length) return false;
  return array.some(function(item) {
    return item.hashMark === hashMark;
  });
};

export const getPercentValue = (current: ?number, total: ?number): number => {
  if (!current || !total) return 0;
  return current / total < 0 ? 0 : current / total > 1 ? 100 : Math.floor((current / total) * 100);
};

export const getHashMark = (): string => {
  return 'lugia' + new Date().getTime() + Math.round(Math.random() * 100000 + 1000);
};

const loop = () => true;

class Upload extends React.Component<UploadProps, StateProps> {
  input: Object;
  static defaultProps = {
    disabled: false,
    areaType: 'default',
    multiple: false,
    showFileList: false,
    limit: Infinity,
    withCredentials: false,
    autoUpload: true,
    method: 'post',
    onProgress: loop,
    onSuccess: loop,
    onComplete: loop,
    onFail: loop,
    onChange: loop,
    defaultTips: {
      uploadText: '上传',
      uploadTips: '请将文件拖到此处',
      failTips: '文件上传失败请重试',
      loadingTips: '文件上传中...',
    },
  };

  static getDerivedStateFromProps(defProps: UploadProps, stateProps: StateProps) {
    const defaultUploadTips = defProps.defaultTips
      ? defProps.defaultTips.uploadTips
      : '请将文件拖到此处';

    if (!stateProps) {
      return {
        classNameStatus: 'default',
        defaultText: defaultUploadTips,
        fileListDone: defProps.fileList || [],
        isAllowUpload: defProps.autoUpload,
      };
    }
    const { classNameStatus, defaultText, fileListDone, isAllowUpload } = stateProps;
    return {
      classNameStatus: 'classNameStatus' in stateProps ? classNameStatus : 'default',
      defaultText: 'defaultText' in stateProps ? defaultText : defaultUploadTips,
      fileListDone: 'fileList' in defProps ? defProps.fileList : fileListDone,
      isAllowUpload: 'isAllowUpload' in stateProps ? isAllowUpload : defProps.autoUpload,
    };
  }

  render() {
    const { themeProps } = this.props;
    return (
      <Container themeProps={themeProps}>
        <GetElement
          {...this.props}
          {...this.state}
          setChoosedFile={this.setChoosedFile}
          setAutoUploadState={this.setAutoUploadState}
          setDeleteList={this.setDeleteList}
        />
      </Container>
    );
  }

  setChoosedFile = (res: Array<Object>): void => {
    const { multiple } = this.props;
    let choosedFiles = res;
    if (!multiple) {
      choosedFiles = [res[0]];
    }
    this.setStateValue(
      {
        choosedFile: choosedFiles,
      },
      () => {
        this.startUpload();
      }
    );
    const { onChange } = this.props;
    onChange && onChange(choosedFiles);
  };

  getChangeUploadState = (typeState: string, name: string, hashMark: string) => {
    let list;
    const { fileListDone } = this.state;
    const newFileList = [...fileListDone];
    if (isIdInArray(hashMark, newFileList)) {
      list = this.updateFieldList(newFileList, hashMark, [{ target: 'status', value: 'loading' }]);
    } else {
      const { areaType } = this.props;
      list = this.appendFileList(newFileList, {
        hashMark,
        name,
        areaType,
        status: typeState,
        percent: 0,
      });
    }
    return {
      classNameStatus: typeState,
      fileListDone: list,
      defaultText: name,
    };
  };

  startUpload = (): void => {
    const { choosedFile = [] } = this.state;
    const len = choosedFile.length;
    if (len <= 0) return;

    const { isAllowUpload } = this.state;
    const { autoUpload, limit = Infinity } = this.props;

    for (let i = 0; i < len; i++) {
      if (i >= limit) break;
      let list;
      const name = choosedFile[i].name;
      const hashMark = choosedFile[i].hashMark || getHashMark();
      choosedFile[i].hashMark = hashMark;
      if (autoUpload || (!autoUpload && isAllowUpload)) {
        list = this.getChangeUploadState('loading', name, hashMark);
        this.setStateValue({ ...list, choosedFile });
      } else {
        list = this.getChangeUploadState('default', name, hashMark);
        this.setStateValue({ ...list, choosedFile });
      }
    }
    if (!autoUpload && !isAllowUpload) return;

    const { url, withCredentials, data, headers, method = 'post', dataType = 'json' } = this.props;
    const dataObject = { url, withCredentials, data, headers, method, dataType };

    for (let i = 0; i < len; i++) {
      if (i >= limit) break;
      this.beforeUpload(dataObject, choosedFile[i], choosedFile[i].hashMark);
      delete choosedFile[i].hashMark;
    }
  };

  beforeUpload = (dataObject: Object, file: Object, hashMark: string) => {
    const { beforeUpload } = this.props;
    if (!beforeUpload) {
      this.startRequest(dataObject, file, hashMark);
    } else {
      beforeUpload(file).then(message => {
        if (message.status) {
          const { accessKey } = this.props;
          const { file } = message;
          if (accessKey) {
            accessKey.forEach(item => {
              dataObject.data[item] = file[item];
              delete file[item];
            });
          }
          this.startRequest(dataObject, file, hashMark);
        }
      });
    }
  };

  startRequest = (dataObject: Object, file: Object, hashMark: string): void => {
    request({
      ...dataObject,
      file,
      onSuccess: res => {
        this.uploadSuccess(res, file, hashMark);
      },
      onFail: res => {
        this.uploadFail(res, hashMark);
      },
      onProgress: res => {
        this.uploadProgress(res, hashMark);
      },
      onComplete: res => {
        this.uploadComplete(res, hashMark);
      },
    });
  };

  getResponse = (res: Array<Object>): ?Array<Object> => {
    if (!res) return;
    const arr = [];
    res.forEach(item => {
      const { areaType, name, percent, status, url } = item;
      arr.push({ areaType, name, percent, status, url });
    });

    return arr;
  };

  uploadSuccess = (res: Object, file: Object, hashMark: string): void => {
    const { fileListDone } = this.state;

    const list = this.updateFieldList(fileListDone, hashMark, [
      { target: 'status', value: 'done' },
      { target: 'url', value: res && res.data && res.data.url },
    ]);
    this.setStateValue({ classNameStatus: 'done', fileListDone: list });
    const { areaType } = this.props;
    if (areaType === 'picture') {
      this.loadPreviewInfo(file);
    }
    this.setStateValue({ isAllowUpload: false });

    const { onSuccess } = this.props;
    onSuccess && onSuccess(res, this.getResponse(fileListDone));
  };

  uploadComplete = (res: Object, hashMark: string): void => {
    const { onComplete } = this.props;
    onComplete && onComplete(res.currentTarget.response);
  };

  uploadProgress = (res: Object, hashMark: string): void => {
    const { loaded, total } = res;
    const percent = getPercentValue(loaded, total);

    const { fileListDone } = this.state;
    const list = this.updateFieldList(fileListDone, hashMark, [
      { target: 'percent', value: percent },
      { target: 'status', value: 'loading' },
    ]);
    this.setStateValue({ classNameStatus: 'loading', fileListDone: list });

    const { onProgress } = this.props;
    onProgress && onProgress({ loaded, total });
  };

  uploadFail = (res: Object, hashMark: string): void => {
    const { fileListDone } = this.state;
    const list = this.updateFieldList(fileListDone, hashMark, [
      { target: 'status', value: 'fail' },
    ]);
    this.setStateValue({ classNameStatus: 'fail', fileListDone: list });

    const { onFail } = this.props;
    onFail && onFail(res);
  };

  updateFieldList = (
    fileListDone: Array<Object>,
    props: string,
    data: Array<Object> = []
  ): Array<Object> => {
    fileListDone.forEach(file => {
      if (file.hashMark === props) {
        data.forEach(item => {
          file[item.target] = item.value;
        });
      }
    });
    return fileListDone;
  };

  appendFileList = (fileListDone: Array<Object>, props: Object): Array<Object> => {
    if (!isEmptyObject(props)) {
      fileListDone.push(props);
    }
    return fileListDone;
  };

  loadPreviewInfo = (file: any): void => {
    const reader = new FileReader();
    reader.readAsDataURL(file);
    reader.onloadend = () => {
      this.setStateValue({ previewUrl: reader.result });
      delete reader.onloadend;
    };
  };

  setDeleteList = (index: number) => {
    const { fileListDone } = this.state;
    fileListDone.splice(index, 1);
    this.setStateValue({ fileListDone });
  };

  setAutoUploadState = (value: boolean) => {
    const { choosedFile } = this.state;
    if (!choosedFile) return;
    this.setStateValue({ isAllowUpload: value }, this.startUpload);
  };

  setStateValue = (props: Object, cbk?: Function) => {
    this.setState(
      {
        ...props,
      },
      () => {
        cbk && cbk();
      }
    );
  };
}

export default Upload;
