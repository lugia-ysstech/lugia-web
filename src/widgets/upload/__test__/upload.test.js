/*
 *@flow
 *
 */

import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import Upload from '../index';
import {
  getIndexInArray,
  getPercentValue,
  isKeyInArray,
  isEmptyObject,
  isIdInArray,
} from '../upload';

import { getFormData, getRequestXHR, getQueryString } from '../request';
import { getIconByType } from '../getelement';
import { getListIconType } from '../../css/upload';
import Enzyme, { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

Enzyme.configure({ adapter: new Adapter() });

describe('Upload Test', () => {
  const themeHocProps = () => true;
  const themeProps = { themeConfig: {}, themeState: {} };
  const getCmp = (target: any): Object => {
    return target.children().instance();
  };
  const target = mount(
    <Upload
      getPartOfThemeHocProps={themeHocProps}
      getPartOfThemeProps={themeHocProps}
      themeProps={themeProps}
      url={'xxxx.test'}
    />
  );
  it('css', () => {
    const target = (
      <Upload
        getPartOfThemeHocProps={themeHocProps}
        getPartOfThemeProps={themeHocProps}
        themeProps={themeProps}
        url={'xxxx.test'}
      />
    );
    expect(renderer.create(target).toJSON()).toMatchSnapshot();
  });

  it('props autoUpload true', () => {
    const target = mount(
      <Upload
        getPartOfThemeHocProps={themeHocProps}
        getPartOfThemeProps={themeHocProps}
        themeProps={themeProps}
        autoUpload={true}
        url={'xxxx.test'}
      />
    );
    expect(getCmp(target).state.isAllowUpload).toEqual(true);
  });

  it('props autoUpload false', () => {
    const target = mount(
      <Upload
        getPartOfThemeHocProps={themeHocProps}
        getPartOfThemeProps={themeHocProps}
        themeProps={themeProps}
        autoUpload={false}
        url={'xxxx.test'}
      />
    );
    expect(getCmp(target).state.isAllowUpload).toEqual(false);
  });

  it('props url', () => {
    const target = mount(
      <Upload
        getPartOfThemeHocProps={themeHocProps}
        getPartOfThemeProps={themeHocProps}
        themeProps={themeProps}
        url={'xxxx.test'}
      />
    );
    expect(target.props().url).toEqual('xxxx.test');
  });

  it('props fileList', () => {
    const target = mount(
      <Upload
        getPartOfThemeHocProps={themeHocProps}
        getPartOfThemeProps={themeHocProps}
        themeProps={themeProps}
        url={'xxxx.test'}
        fileList={[
          { id: 1, name: '文件11111.jpg', status: 'done' },
          { id: 2, name: '文件666.doc', status: 'fail' },
        ]}
      />
    );
    expect(getCmp(target).state.fileListDone).toEqual([
      { id: 1, name: '文件11111.jpg', status: 'done' },
      { id: 2, name: '文件666.doc', status: 'fail' },
    ]);
  });

  it('props fileList null', () => {
    const target = mount(
      <Upload
        getPartOfThemeHocProps={themeHocProps}
        getPartOfThemeProps={themeHocProps}
        themeProps={themeProps}
        url={'xxxx.test'}
      />
    );
    expect(getCmp(target).state.fileListDone).toEqual([]);
  });

  function checkFindIndex(data: Array<string>, key: string, expectation: number) {
    it('Function getIndexInArray ', () => {
      const res = getIndexInArray(data, key.toLowerCase());
      expect(res).toEqual(expectation);
    });
  }

  checkFindIndex(['jpg', 'png', 'jpeg', 'gif', 'svg', 'bmp'], 'jpg', 0);
  checkFindIndex(['jpg', 'png', 'jpeg', 'gif', 'svg', 'bmp'], 'jPg', 0);
  checkFindIndex(['jpg', 'png', 'jpeg', 'gif', 'svg', 'bmp'], 'JPG', 0);
  checkFindIndex(['jpg', 'png', 'jpeg', 'gif', 'svg', 'bmp'], 'jpEg', 2);
  checkFindIndex(['jpg', 'png', 'jpeg', 'gif', 'svg', 'bmp'], 'ABC', -1);
  checkFindIndex(['jpg', 'png', 'jpeg', 'gif', 'svg', 'bmp'], 'BMP', 5);

  function checkKeyInArr(data: Array<string>, key: string, expectation: boolean) {
    it('Function isKeyInArr ', () => {
      const res = isKeyInArray(data, key.toLowerCase());
      expect(res).toEqual(expectation);
    });
  }

  checkKeyInArr(['jpg', 'png', 'jpeg', 'gif', 'svg', 'bmp'], 'jpg', true);
  checkKeyInArr(['jpg', 'png', 'jpeg', 'gif', 'svg', 'bmp'], 'jPg', true);
  checkKeyInArr(['jpg', 'png', 'jpeg', 'gif', 'svg', 'bmp'], 'JPG', true);
  checkKeyInArr(['jpg', 'png', 'jpeg', 'gif', 'svg', 'bmp'], 'jpEg', true);
  checkKeyInArr(['jpg', 'png', 'jpeg', 'gif', 'svg', 'bmp'], 'ABC', false);
  checkKeyInArr(['jpg', 'png', 'jpeg', 'gif', 'svg', 'bmp'], 'BMP', true);

  function checkGetPercentValue(current: ?number, total: ?number, expectation: number) {
    it('Function getPercentValue ', () => {
      const res = getPercentValue(current, total);
      expect(res).toEqual(expectation);
    });
  }

  checkGetPercentValue(2345, 3455, 67);
  checkGetPercentValue(null, 3455, 0);
  checkGetPercentValue(2345, undefined, 0);
  checkGetPercentValue(3488, 3455, 100);

  function checkGetListIconType(fileName: ?string, expectation: string) {
    it('Function getListIconType ', () => {
      const res = getListIconType(fileName);
      expect(res).toEqual(expectation);
    });
  }

  checkGetListIconType('http://jshdfj.jpg', 'picture');
  checkGetListIconType('http://jshdfj.png', 'picture');
  checkGetListIconType('http://jshdfj.JPEG', 'picture');
  checkGetListIconType('http://jshdfj.avi', 'video');
  checkGetListIconType('http://jshdfj.MP4', 'video');
  checkGetListIconType('http://jshdfj.cccc', 'file');
  checkGetListIconType('http://jshdfj.text', 'file');
  checkGetListIconType('http://jshdfj.doc', 'file');
  checkGetListIconType('', 'file');
  checkGetListIconType(null, 'file');
  checkGetListIconType(undefined, 'file');

  it('Function getRequestXHR ', () => {
    const res = getRequestXHR();
    expect(res instanceof window.XMLHttpRequest).toEqual(true);
  });

  function checkGetStringFromObject(data: ?Object, expectation: string) {
    it('Function getQueryString ', () => {
      const res = getQueryString(data);
      expect(res).toEqual(expectation);
    });
  }

  checkGetStringFromObject({ u: 23, name: 'lll', file: 'xsdss' }, 'u=23&name=lll&file=xsdss');
  checkGetStringFromObject({ cc: 66, df: 'lll', file: 'xsdss' }, 'cc=66&df=lll&file=xsdss');
  checkGetStringFromObject(null, '');
  checkGetStringFromObject(undefined, '');
  checkGetStringFromObject({}, '');
  checkGetStringFromObject({ hello: 'world' }, 'hello=world');

  function checkGetParamsData(data: Object) {
    it('Function getFormData ', () => {
      const res = getFormData(data, {});
      expect(res instanceof FormData).toEqual(true);
    });
  }

  checkGetParamsData({ data: { a: 123, b: 223 }, name: 'file', file: '666' });
  checkGetParamsData({ data: { a: 777, b: 888 }, name: 'abc', file: '567' });

  function checkGetIconByType(status: ?string, expectation: ?string, props?: Object = {}) {
    it('Function GetIconByType ', () => {
      const res = getIconByType(
        {
          defaultTips: {
            uploadText: '上传',
            uploadTips: '请将文件拖到此处',
            failTips: '文件上传失败请重试',
            loadingTips: '文件上传中...',
          },
        },
        status,
        props
      );
      const { type } = props;
      if (!status || !res) {
        expect(res).toBe(expectation);
      } else if (typeof res === 'string' || (type === 1 && status === 'default')) {
        expect(res).toEqual(expectation);
      } else {
        const { iconClass } = res.props;
        expect(iconClass).toBe(expectation);
      }
    });
  }

  checkGetIconByType('default', '上传', { type: 1 });
  checkGetIconByType('default', 'lugia-icon-financial_upload right');
  checkGetIconByType('loading', 'lugia-icon-financial_loading_o loading');
  checkGetIconByType(undefined, null);
  checkGetIconByType(undefined, null, { type: 1 });

  function setStateValue(props: Object, expectation: Object) {
    it('Function setStateValue ', () => {
      getCmp(target).setStateValue(props);
      expect(getCmp(target).state).toEqual(expectation);
    });
  }

  setStateValue(
    { classNameStatus: 'done', defaultText: '文件已上传成功！' },
    {
      classNameStatus: 'done',
      defaultText: '文件已上传成功！',
      fileListDone: [],
      isAllowUpload: true,
    }
  );
  setStateValue(
    { classNameStatus: 'loading', defaultText: '文件正在上传！' },
    {
      classNameStatus: 'loading',
      defaultText: '文件正在上传！',
      fileListDone: [],
      isAllowUpload: true,
    }
  );
  setStateValue(
    { classNameStatus: 'default', defaultText: '文件已上传成功！', abc: 'abc' },
    {
      abc: 'abc',
      classNameStatus: 'default',
      defaultText: '文件已上传成功！',
      fileListDone: [],
      isAllowUpload: true,
    }
  );

  function checkGetFileList(
    fileListDone: Array<Object>,
    props: number,
    expectation: Array<Object>,
    data: Array<Object>
  ) {
    it('Function updateFieldList ', () => {
      const res = getCmp(target).updateFieldList(fileListDone, props, data);
      expect(res).toEqual(expectation);
    });
  }

  checkGetFileList(
    [
      { hashMark: 1, name: '文件11111.jpg', status: 'loading', percent: 0 },
      { hashMark: 2, name: '文件22222222.jpg', status: 'loading' },
    ],
    1,
    [
      { hashMark: 1, name: '文件11111.jpg', status: 'loading', percent: 20 },
      { hashMark: 2, name: '文件22222222.jpg', status: 'loading' },
    ],
    [{ target: 'percent', value: 20 }]
  );
  checkGetFileList(
    [
      { id: 1, name: '文件11111.jpg', status: 'loading', percent: 20 },
      { id: 2, name: '文件22222222.jpg', status: 'loading' },
    ],
    3,
    [
      { id: 1, name: '文件11111.jpg', status: 'loading', percent: 20 },
      { id: 2, name: '文件22222222.jpg', status: 'loading' },
    ],
    [{ target: 'percent', value: 50 }]
  );

  function checkIsEmptyObject(data: Object, expectation: boolean) {
    it('Function isEmptyObject ', () => {
      const res = isEmptyObject(data);
      expect(res).toEqual(expectation);
    });
  }

  checkIsEmptyObject({}, true);
  checkIsEmptyObject({ aa: 1 }, false);

  function checkAppendFileList(
    fileListDone: Array<Object>,
    props: Object,
    expectation: Array<Object>
  ) {
    it('Function updateFieldList ', () => {
      const res = getCmp(target).appendFileList(fileListDone, props);
      expect(res).toEqual(expectation);
    });
  }

  checkAppendFileList(
    [{ hashMark: 1, name: '文件11111.jpg', status: 'loading', percent: 0 }],
    { hashMark: 2, name: '文件2222.jpg', status: 'loading' },
    [
      { hashMark: 1, name: '文件11111.jpg', status: 'loading', percent: 0 },
      { hashMark: 2, name: '文件2222.jpg', status: 'loading' },
    ]
  );
  checkAppendFileList([{ hashMark: 1, name: '文件11111.jpg', status: 'loading', percent: 0 }], {}, [
    { hashMark: 1, name: '文件11111.jpg', status: 'loading', percent: 0 },
  ]);

  function uploadProgress(props: Object, expectation: string, expectation2: Array<Object>) {
    it('Function uploadProgress ', () => {
      getCmp(target).setStateValue({
        fileListDone: [{ hashMark: 1, name: '文件11111.jpg', status: 'default' }],
      });
      getCmp(target).uploadProgress(props, 1);
      expect(getCmp(target).state.classNameStatus).toEqual(expectation);
      expect(getCmp(target).state.fileListDone).toEqual(expectation2);
    });
  }

  uploadProgress({ loaded: 0, total: 2048 }, 'loading', [
    { hashMark: 1, name: '文件11111.jpg', status: 'loading', percent: 0 },
  ]);
  uploadProgress({ loaded: 1024, total: 2048 }, 'loading', [
    { hashMark: 1, name: '文件11111.jpg', status: 'loading', percent: 50 },
  ]);
  uploadProgress({ loaded: 2048, total: 2048 }, 'loading', [
    { hashMark: 1, name: '文件11111.jpg', status: 'loading', percent: 100 },
  ]);
  uploadProgress({}, 'loading', [
    { hashMark: 1, name: '文件11111.jpg', status: 'loading', percent: 0 },
  ]);

  function uploadSuccess(props: Object, expectation: string, expectation2: Array<Object>) {
    it('Function uploadSuccess ', () => {
      getCmp(target).setStateValue({
        fileListDone: [{ hashMark: 1, name: '文件11111.jpg', status: 'default' }],
      });
      getCmp(target).uploadSuccess(props, files, 1);
      expect(getCmp(target).state.classNameStatus).toEqual(expectation);
      expect(getCmp(target).state.fileListDone).toEqual(expectation2);
    });
  }

  uploadSuccess({ data: { url: 'test.jpg' } }, 'done', [
    { hashMark: 1, name: '文件11111.jpg', status: 'done', url: 'test.jpg' },
  ]);

  function checkisIdInArray(id: string, array: Array<Object>, expectation: boolean) {
    it('Function isIdInArray ', () => {
      const res = isIdInArray(id, array);
      expect(res).toEqual(expectation);
    });
  }

  checkisIdInArray('1', [{ hashMark: '1', name: '文件11111.jpg', status: 'default' }], true);
  checkisIdInArray('2', [{ hashMark: '1', name: '文件11111.jpg', status: 'default' }], false);
  checkisIdInArray('1', [], false);

  const files = [
    {
      lastModified: 1541066150245,
      lastModifiedDate: new Date(1541066150245),
      name: 'headPic.jpg',
      size: 53021,
      type: 'image/jpeg',
      webkitRelativePath: '',
    },
  ];

  function setAutoUploadState(
    props: boolean,
    choosedFileList: ?Array<Object>,
    expectation: boolean
  ) {
    it('Function setAutoUploadState ', () => {
      const target = mount(
        <Upload
          getPartOfThemeHocProps={themeHocProps}
          getPartOfThemeProps={themeHocProps}
          themeProps={themeProps}
          url={'xxxx.test'}
        />
      );
      choosedFileList && getCmp(target).setStateValue({ choosedFile: choosedFileList });
      getCmp(target).setAutoUploadState(props);
      expect(getCmp(target).state.isAllowUpload).toEqual(expectation);
    });
  }

  setAutoUploadState(false, undefined, true);
  setAutoUploadState(true, files, true);
  setAutoUploadState(false, files, false);

  function setDeleteList(index: number, expectation: Array<Object>) {
    it('Function setDeleteList ', () => {
      getCmp(target).setStateValue({
        fileListDone: [
          { hashMark: 1, name: '文件11111.jpg', status: 'done' },
          { hashMark: 2, name: '文件2222.jpg', status: 'default' },
        ],
      });
      getCmp(target).setDeleteList(index);
      expect(getCmp(target).state.fileListDone).toEqual(expectation);
    });
  }

  setDeleteList(1, [{ hashMark: 1, name: '文件11111.jpg', status: 'done' }]);
  setDeleteList(0, [{ hashMark: 2, name: '文件2222.jpg', status: 'default' }]);
  setDeleteList(2, [
    { hashMark: 1, name: '文件11111.jpg', status: 'done' },
    { hashMark: 2, name: '文件2222.jpg', status: 'default' },
  ]);

  function checkUploadFail(
    props: Object,
    id: number,
    expectation: string,
    expectation2: Array<Object>
  ) {
    it('Function checkUploadFail ', () => {
      getCmp(target).setStateValue({
        fileListDone: [{ hashMark: 1, name: '文件11111.jpg', status: 'loading' }],
      });
      getCmp(target).uploadFail(props, id);
      expect(getCmp(target).state.classNameStatus).toEqual(expectation);
      expect(getCmp(target).state.fileListDone).toEqual(expectation2);
    });
  }

  checkUploadFail({}, 1, 'fail', [{ hashMark: 1, name: '文件11111.jpg', status: 'fail' }]);
  checkUploadFail({}, 2, 'fail', [{ hashMark: 1, name: '文件11111.jpg', status: 'loading' }]);

  function checkSetChoosedFile(expectation: Array<Object>) {
    it('Function checkSetChoosedFile ', () => {
      getCmp(target).setChoosedFile(files);
      expect(getCmp(target).state.choosedFile).toEqual(expectation);
    });
  }

  checkSetChoosedFile(files);
});
