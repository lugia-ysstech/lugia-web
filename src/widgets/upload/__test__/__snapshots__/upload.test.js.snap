// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload Test css 1`] = `
.c4 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  display: inline-block;
}

.c4.right {
  position: absolute;
  right: 12px;
  top: 50%;
  -webkit-transform: translate(0,-50%);
  -ms-transform: translate(0,-50%);
  transform: translate(0,-50%);
}

.c4.delete {
  display: none;
}

.c4.areaIcon {
  font-size: 36px;
}

.c4.icon-mark {
  margin: 0 6px 0 0;
}

.c4.loading {
  -webkit-animation: iECmZH 0.8s linear infinite;
  animation: iECmZH 0.8s linear infinite;
  margin: 0 12px 0 0;
}

.c2 {
  width: 100%;
  display: none;
}

.c1 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  box-sizing: border-box;
}

.c3 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
  box-sizing: border-box;
  position: relative;
}

.c0 {
  position: relative;
}

<div
  className="c0"
  style={Object {}}
>
  <div
    className="c1"
    style={Object {}}
  >
    <input
      accept="*"
      className="c2"
      disabled={false}
      id="upload"
      multiple={false}
      onChange={[Function]}
      style={Object {}}
      type="file"
    />
    <div
      className="c3"
      disabled={false}
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "borderBottomColor": "#d0d3da",
          "borderBottomLeftRadius": "0.4rem",
          "borderBottomRightRadius": "0.4rem",
          "borderBottomStyle": "solid",
          "borderBottomWidth": "0.1rem",
          "borderLeftColor": "#d0d3da",
          "borderLeftStyle": "solid",
          "borderLeftWidth": "0.1rem",
          "borderRightColor": "#d0d3da",
          "borderRightStyle": "solid",
          "borderRightWidth": "0.1rem",
          "borderTopColor": "#d0d3da",
          "borderTopLeftRadius": "0.4rem",
          "borderTopRightRadius": "0.4rem",
          "borderTopStyle": "solid",
          "borderTopWidth": "0.1rem",
          "color": "#c2c5ce",
          "fontSize": "1.4rem",
          "height": "3rem",
          "paddingBottom": "0rem",
          "paddingLeft": "1.2rem",
          "paddingRight": "0rem",
          "paddingTop": "0rem",
        }
      }
    >
      <i
        className="lugia-icon-financial_upload right   c4"
        disabled={false}
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        style={Object {}}
      />
      请将文件拖到此处
    </div>
  </div>
</div>
`;
